### PR TITLE
86cuewzcv/trip booking page

### PR DESCRIPTION
**Pull request by RJR-003**
### Description
This pull request includes the booking page functionality and its language translation files.

### Changes
- Implement booking functionality for the trips
- Implementing necessary toast messages for the POST API.
- Implementing rerouting when no booking data is found.
- Implement language persistence even when reloading.
- Implementing a custom confirmation dialog box for booking page.
- Implementing responsiveness.
- Dynamic data and form loading along with fare calculations in the booking page.

### Important note:
Review and merge this branch after the merging of https://github.com/nijin-26/bus-ticket-booking-react/pull/30 pull request.

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -10,12 +10,11 @@ import { ThemeProvider as CustomThemeProvider } from '@emotion/react';
 import { CssBaseline, ThemeProvider as MuiThemeProvider } from '@mui/material';
 import { createTheme } from '@mui/material';
 
-//React-toastify
+import { useAppSelector } from './app/hooks';
+// import { getMyBookings } from './api/endpoints/ticket.api';
```
Comment (abhijith-biju) [inline src/App.tsx]: remove commented code.

File: src/components/FareDetails/FareDetails.tsx [hunk 2]
```diff
@@ -7,25 +7,28 @@ interface IFareDetailsProps {
...
     return (
         <Stack direction={'row'} spacing={1}>
             <Typography
                 variant="body1"
-                fontSize={font.h2}
+                fontSize={isMinWidth ? font.h2 : font.md}
                 fontWeight={font.fontWeightMedium}
             >
                 {t('checkoutTitle')}
                 {'  '}:
```
Comment (akshay-krishna) [inline src/components/FareDetails/FareDetails.tsx:27]: is this necessary

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx [hunk 3]
```diff
@@ -3,40 +3,70 @@ import { Box, Grid, MenuItem, Paper, Typography } from '@mui/material';
...
+            navigate(`${paths.ticket}/${responseBook.pnrNumber}`, {
+                state: responseBook,
+                replace: true,
+            });
+        } catch (error) {
+            if (axios.isAxiosError<IAuthResponseError>(error)) {
+                const errorMessage = error.response?.data.message;
+
+                if (errorMessage === 'Required seat already booked') {
+                    toast.error(t('seatAlreadyBookedErrorMessage'));
```
Comment (akshay-krishna) [inline src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx:52]: better to keep toast Id

File: src/components/FareDetails/FareDetails.tsx (hunk 2 above)
Comment (RJR-003) [reply src/components/FareDetails/FareDetails.tsx:27]: yes it is, it is for providing a space for the fare calculation part

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx [hunk 4]
```diff
@@ -3,40 +3,70 @@ import { Box, Grid, MenuItem, Paper, Typography } from '@mui/material';
...
 
     const [languageChangeKey, setLanguageChangeKey] = useState(0);
-    const formikRef = useRef<FormikProps<IPassengerDetails>>(null);
-    const navigate = useNavigate();
 
+    const postBookingData = async (obj: IPassengersInputFromFormik) => {
+        const inputObj = convertFormikDataToApiData(obj);
+        try {
+            const responseBook = await bookTicket(tripId, inputObj);
+            console.log(responseBook);
```
Comment (abhijith-biju) [inline src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx]: remove console logs

**Review by abhijith-biju — APPROVED** [review]

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx (hunk 3 above)
Comment (RJR-003) [reply src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx:52]: okay

**Review by akshay-krishna — APPROVED** [review]

File: src/app/features/languageSlice.ts [hunk 5]
```diff
@@ -6,10 +6,12 @@ interface ILanguageState {
     currentLanguage: LanguageCode;
 }
 
+const localStorageLanguage = localStorage.getItem('language');
```
Comment (lekshmi615213) [inline src/app/features/languageSlice.ts]: can we rename this variable ?localStorageLanguage

File: src/app/features/languageSlice.ts [hunk 6]
```diff
@@ -6,10 +6,12 @@ interface ILanguageState {
...
 const initialState: ILanguageState = {
-    currentLanguage: i18n.language as LanguageCode.English,
+    currentLanguage: localStorageLanguage
+        ? (localStorageLanguage as LanguageCode)
+        : (i18n.language as LanguageCode.English),
```
Comment (lekshmi615213) [inline src/app/features/languageSlice.ts]: casting needed? we could use type gurard

File: src/components/FareDetails/FareDetails.tsx [hunk 7]
```diff
@@ -7,25 +7,28 @@ interface IFareDetailsProps {
...
+    const { t } = useTranslation('tripDetails');
     const totalFare = noOfSeats * farePerSeat;
-    const totalFareInString = `${noOfSeats} seat${
+    const totalFareInString = `${noOfSeats} ${t('seat')}${
         noOfSeats > 1 ? 's' : ''
     } x ₹${farePerSeat} = ₹${totalFare}`;
 
-    const { t } = useTranslation('tripDetails');
     const { font } = useTheme();
+    const isMinWidth = useMediaQuery('(min-width:600px)');
```
Comment (lekshmi615213) [inline src/components/FareDetails/FareDetails.tsx]: 600 can be magi cnumbers

File: src/app/features/languageSlice.ts [hunk 8]
```diff
@@ -19,6 +21,8 @@ const languageSlice = createSlice({
             i18n.changeLanguage(action.payload).catch((error) => {
                 console.error('Something went wrong loading', error);
```
Comment (lekshmi615213) [inline src/app/features/languageSlice.ts]: would be great to show some message to user (unknown error or something similar) for all catch blocks

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx [hunk 9]
```diff
@@ -3,40 +3,70 @@ import { Box, Grid, MenuItem, Paper, Typography } from '@mui/material';
...
     const theme = useTheme();
 
     const [languageChangeKey, setLanguageChangeKey] = useState(0);
-    const formikRef = useRef<FormikProps<IPassengerDetails>>(null);
-    const navigate = useNavigate();
 
+    const postBookingData = async (obj: IPassengersInputFromFormik) => {
+        const inputObj = convertFormikDataToApiData(obj);
+        try {
+            const responseBook = await bookTicket(tripId, inputObj);
```
Comment (lekshmi615213) [inline src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx]: meaningful variable name?

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx [hunk 10]
```diff
@@ -3,40 +3,70 @@ import { Box, Grid, MenuItem, Paper, Typography } from '@mui/material';
...
+                const errorMessage = error.response?.data.message;
+
+                if (errorMessage === 'Required seat already booked') {
+                    toast.error(t('seatAlreadyBookedErrorMessage'));
+                    navigate(paths.home, { replace: true });
+                }
+            } else
+                toast.error(t('apiErrorMessage'), { toastId: 'apiErrorMsg' });
+        } finally {
+            loaderFun(false);
```
Comment (lekshmi615213) [inline src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx]: loaderFunction ?

File: src/pages/TripBookingPage/TripBookingPage.tsx [hunk 11]
```diff
@@ -1,54 +1,129 @@
...
+    const [showDialog, setShowDialog] = useState(false);
+    const [showLoader, setShowLoader] = useState(false);
+    const setShowLoaderFun = (bool: boolean) => {
+        setShowLoader(bool);
+    };
+    const formikRef = useRef<FormikProps<IPassengersInputFromFormik>>(null);
+
+    const handleFormSubmit = () => {
+        if (formikRef.current?.isValid) {
+            setShowDialog(true);
```
Comment (lekshmi615213) [inline src/pages/TripBookingPage/TripBookingPage.tsx:44]: (formikRef.current?.isValid) is it the valid case or invalid case?

File: src/pages/TripBookingPage/TripBookingPage.tsx [hunk 12]
```diff
@@ -1,54 +1,129 @@
...
+        } else {
+            formikRef.current?.handleSubmit(); // to retrigger validation of the form
+        }
+    };
+    useEffect(() => {
+        if (state.seats.length === 0) {
+            navigate(-1);
+            toast.error(t('absentStateError'));
```
Comment (lekshmi615213) [inline src/pages/TripBookingPage/TripBookingPage.tsx:52]: noAvailableSeatError ?

File: src/app/features/languageSlice.ts (hunk 5 above)
Comment (RJR-003) [reply src/app/features/languageSlice.ts]: Okay

File: src/app/features/languageSlice.ts (hunk 8 above)
Comment (RJR-003) [reply src/app/features/languageSlice.ts]: okay, i will add a toast

File: src/components/FareDetails/FareDetails.tsx (hunk 7 above)
Comment (RJR-003) [reply src/components/FareDetails/FareDetails.tsx]: okay, i will change that

File: src/app/features/languageSlice.ts (hunk 6 above)
Comment (RJR-003) [reply src/app/features/languageSlice.ts]: Okay i will change the type assertion.

File: src/app/features/languageSlice.ts (hunk 5 above)
Comment (RJR-003) [reply src/app/features/languageSlice.ts]: Yes i will change it

File: src/app/features/languageSlice.ts (hunk 6 above)
Comment (RJR-003) [reply src/app/features/languageSlice.ts]: Okay, I have removed the type assertions

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx (hunk 9 above)
Comment (RJR-003) [reply src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx]: I have changed it.

File: src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx (hunk 10 above)
Comment (RJR-003) [reply src/pages/TripBookingPage/PassengerDetailsForm/PassengerDetailsForm.tsx]: I will add that

File: src/pages/TripBookingPage/TripBookingPage.tsx (hunk 11 above)
Comment (RJR-003) [reply src/pages/TripBookingPage/TripBookingPage.tsx:44]: valid case. It is the case where all the validation checks have passed and a confirmation modal is shown to confirm the booking.

File: src/pages/TripBookingPage/TripBookingPage.tsx (hunk 12 above)
Comment (RJR-003) [reply src/pages/TripBookingPage/TripBookingPage.tsx:52]: This is the condition when the store doesn't have the total current booking data. This happens when the user reloads the booking page or directly navigates to the booking page without selecting any seats.

**Pull request merged**